swarm/pss/pssapi.go

```diff
@@ -90,48 +103,77 @@ func (pssapi *API) GetPublicKey() []byte {
 }
 
 // Set Public key to associate with a particular Pss peer
-func (pssapi *API) SetPeerPublicKey(addr []byte, topic whisper.TopicType, pubkey []byte) error {
-	var potaddr pot.Address
-	copy(potaddr[:], addr)
-	pssapi.Pss.SetPeerPublicKey(potaddr, topic, crypto.ToECDSAPub(pubkey))
+func (pssapi *API) SetPeerPublicKey(pubkey []byte, topic whisper.TopicType, addr PssAddress) error {
+	pssapi.Pss.SetPeerPublicKey(crypto.ToECDSAPub(pubkey), topic, addr)
 	return nil
 }
 
+// Set Symmetric key for communication with a particular Pss peer
+//
+// if the addtocache parameter is set to false, pss will add this key to the collection of keys used to decrypt incoming messages.
+//
+// returns the string id auto generated by whisper. Use this to retrieve the key later with GetSymKey
+func (pssapi *API) SetPeerSymmetricKey(symkey []byte, topic whisper.TopicType, addr PssAddress, expires time.Duration, addtocache bool) (string, error) {
+	return pssapi.Pss.SetSymmetricKey(symkey, topic, addr, expires, addtocache)
+}
+
 // Generate a new symkey for a Pss peer, and send requesting a key in return
 // The addresslength parameter decides how many bytes of the address to reveal in transit. -1 equals showing all. 0 means show none.
-func (pssapi *API) Handshake(to []byte, topic whisper.TopicType, addresslength int8) (string, error) {
-	return pssapi.sendKey(to, &topic, int(addresslength))
+func (pssapi *API) Handshake(pubkey string, topic whisper.TopicType, to PssAddress) (string, error) {
+	return pssapi.sendKey(common.FromHex(pubkey), &topic, "", to)
+}
+
+// Returns incoming and outgoing symkey pair for a particular Pss peer
+func (self *API) MatchSymKey(key string) (string, error) {
+	id := self.Pss.symKeyPairIndex[key]
+	if id == nil {
+		return "", fmt.Errorf("invalid key id '%s'", key)
+	}
+	return *id, nil
 }
 
 // PssAPITest are temporary API calls for development use only
-//
-// These symbols should NOT be included in production environment
 type APITest struct {
 	*Pss
 }
 
-// Only for debugging and tests!
 func NewAPITest(ps *Pss) *APITest {
 	return &APITest{Pss: ps}
 }
 
-// Returns incoming and outgoing symkey pair for a particular Pss peer
-func (self *APITest) GetSymKeys(to []byte, topic whisper.TopicType) ([]byte, error) {
-	var potaddr pot.Address
-	copy(potaddr[:], to)
-	sendsymkey, err := self.w.GetSymKey(self.peerPool[potaddr][topic].sendsymkey)
+// get symkey matching symkeyid in pss symkeypool
+func (self *APITest) GetSymKey(keyid string) ([]byte, error) {
+	return self.Pss.w.GetSymKey(keyid)
+}
+
+// get all symkey ids in pss symkeypool
+func (self *APITest) GetSymKeys() (ids []string, err error) {
+	for id, _ := range self.Pss.symKeyPool {
+		ids = append(ids, id)
+	}
+	return ids, nil
+}
+
+// manually set in- and outgoing pair of symkeys
+// equals state after dh symkey exchange (handshake)
+func (self *APITest) SetSymKeys(recvkey []byte, sendkey []byte, topic whisper.TopicType, addr PssAddress) (keyids [2]string, err error) {
```

Comment:
this [2] seems convoluted. 
Could we not have just `(keyid0, keyid1 string, err error)` as the return values?

Reply:
return value from an api call must be one data structure encapsulating all returned data, it seems.

Reply:
yes gotcha.